publisher-sdk/src/main/java/com/criteo/publisher/LiveCdbCallListener.java

```diff
@@ -0,0 +1,131 @@
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+package com.criteo.publisher;
+
+import androidx.annotation.NonNull;
+import com.criteo.publisher.annotation.Internal;
+import com.criteo.publisher.bid.BidLifecycleListener;
+import com.criteo.publisher.model.CacheAdUnit;
+import com.criteo.publisher.model.CdbRequest;
+import com.criteo.publisher.model.CdbResponse;
+import com.criteo.publisher.model.CdbResponseSlot;
+import com.criteo.publisher.util.PreconditionsUtil;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Implementation specific to listening Cdb calls for serving a live request
+ */
+@Internal
+public class LiveCdbCallListener extends CdbCallListener {
+
+  private final BidListener bidListener;
+  private final BidManager bidManager;
+  private final CacheAdUnit cacheAdUnit;
+  private final BidLifecycleListener bidLifecycleListener;
+  private final AtomicBoolean isListenerTriggered = new AtomicBoolean(false);
+
+  public LiveCdbCallListener(
+      @NonNull BidListener bidListener,
+      @NonNull BidLifecycleListener bidLifecycleListener,
+      @NonNull BidManager bidManager,
+      @NonNull CacheAdUnit cacheAdUnit
+  ) {
+    super(bidLifecycleListener, bidManager);
+    this.bidListener = bidListener;
+    this.bidLifecycleListener = bidLifecycleListener;
+    this.bidManager = bidManager;
+    this.cacheAdUnit = cacheAdUnit;
+  }
+
+  /**
+   * Triggered when a response is fetched before the expiration of the network timeout. Following
+   * this, two things can happen:
+   * <p>
+   * 1. If {@link LiveCdbCallListener#onTimeBudgetExceeded()}  hasn't been triggered yet (on a
+   * separate thread), a bid response is returned to the caller via {@link
+   * BidListener#onBidResponse(CdbResponseSlot)} unless the {@link CdbResponseSlot} has been
+   * silenced, in which case {@link BidListener#onNoBid()} is triggered instead.
+   * <p>
+   * 2. If {@link LiveCdbCallListener#onTimeBudgetExceeded()} has been triggered, then either {@link
+   * BidListener#onNoBid()} or {@link BidListener#onBidResponse(CdbResponseSlot)} were already
+   * triggered. The only action that needs to be taken here is to cache the {@link
+   * CdbResponseSlot}.
+   */
+  @Override
+  public void onCdbResponse(
+      @NotNull CdbRequest cdbRequest,
+      @NotNull CdbResponse cdbResponse
+  ) {
+    super.onCdbResponse(cdbRequest, cdbResponse);
+
+    if (cdbResponse.getSlots().size() > 1) {
+      PreconditionsUtil.throwOrLog(new IllegalStateException(
+          "During a live request, only one bid will be fetched at a time."));
+    }
+    if (isListenerTriggered.compareAndSet(false, true)) {
+      if ((cdbResponse.getSlots().size() == 1) && !bidManager.isBidSilent(cdbResponse.getSlots()
```

Comment:
Why checking there if slot is silenced? I thought the intent was to prevent CDB call?
There it just prevents passing a valid response to listener, isn't?

Reply:
The bid silence logic will be handled in a separate commit.